scripts/voucherHubUtils.ts

```diff
@@ -39,58 +40,56 @@ export async function upgradeProxy(
-    return await ethers.provider.getStorage(voucherHubAddress, codeHashSlot);
-}
-
-/**
- * Get a hardcoded value of the VoucherProxy creationCode hash to prevent
- * botched upgrades.
- *
- * @param voucherBeaconAddress
- * @returns value of the hash
- */
-export async function getVoucherProxyCreationCodeHash(voucherBeaconAddress: string) {
-    const factory = await ethers.getContractFactory('VoucherProxy');
-    const tx = await factory.getDeployTransaction(voucherBeaconAddress);
-    /**
-     * tx.data is the same as Solidity value of:
-     * ```
-     * abi.encodePacked(
-     *     type(VoucherProxy).creationCode, // bytecode
-     *     abi.encode($._voucherBeacon) // constructor args
-     * )
-     * ```
-     */
-    return ethers.keccak256(tx.data);
-
-    // TODO comment the implementation and return a hardcoded hash.
-    // For some reason, the hash is different between "hardhat test" and "hardhat coverage".
-    // '0x6b4bda6ca928b9724d26ee10eb17168dcd9c632e1905c854c10b78a05cd83398' // hardhat test
-    // '0x97822cb09c6322b4ccd1c4bb70005e4a3ce60099d392676696ce3d3d69e8949f' // hardhat coverage
-
-    // console.log(ethers.keccak256(tx.data));
-    // return '<hash>'
+export async function getExpectedVoucherProxyCodeHash(voucherBeaconAddress: string) {
+    const chainId = (await ethers.provider.getNetwork()).chainId.toString();
+    const config = await getDeploymentConfig(Number(chainId));
+    if (!config.factory || (hre as any).__SOLIDITY_COVERAGE_RUNNING) {
+        /**
+         * @dev Voucher proxy code hash is different from the production one:
+         * - when running "test" without generic factory since voucher beacon address
+         *   becomes not-deterministic (the EOA deployer nonce is involved).
+         * - when running "coverage" since the later compiles contracts with altered
+         *   solc options.
+         *
+         * The expected returned value below mimics the following Solidity piece of code:
+         * ```
+         * abi.encodePacked(
+         *     type(VoucherProxy).creationCode, // bytecode
+         *     abi.encode($._voucherBeacon) // constructor args
+         * )
+         * ```
+         */
+        return ethers.solidityPackedKeccak256(
+            ['bytes', 'bytes32'],
+            [
+                VoucherProxy__factory.bytecode, // bytecode
+                ethers.zeroPadValue(voucherBeaconAddress, 32), // constructor args
+            ],
+        );
+    } else {
+        /**
+         * @dev This is the voucher proxy code hash value expected in production,
+         * where contracts are deployed through a generic factory, hence having
+         * deterministic addresses.
+         *
+         * Note: Look very carefully before updating this value to avoid messing with
+         * existing vouchers already deployed in production.
+         */
+        return '0xebe0c40d5e931d865b61665b4e9435cc65ece10955457af1b4271e3438e4949e';
```

Comment:
How did you compute this

Reply:
It's actually just the value we get at this time..! (but we could compute it with the formula above)

Reply:
The original need was to guaranty an hardcoded hash value that should not change on future upgrades, but we need to handle other cases too (no factory, coverage, ..)